### PR TITLE
Add support for `QNode` mid-circuit measurement config arguments

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
As name says. `QNode`s accept new arguments for configuring mid-circuit measurements (`mcm_method` and `postselect_mode`). This PR adds support for specifying `mcm_method` and using is correctly with `qjit`.

**Description of the Change:**
* Use `mcm_method` to dispatch to correct transform for mid-circuit measurements.
* Propagate `postselect_mode` to `MidCircuitMeasure.trace_quantum` to determine whether or not to use hardware like behaviour
* Changed `trace_quantum_tape` to `trace_quantum_operations`.
* Tests
* Changelog entry

Couple of details:
* Setting `mcm_method="one-shot"` will _always_ trigger the use of `dynamic_one_shot`, regardless of whether or not the circuit has mid-circuit measurements.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mudit2812** [comment]
[sc-64129]

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 1]
```diff
@@ -240,9 +294,10 @@ def circuit(x):
     @pytest.mark.parametrize(
         "meas_obj", [qml.PauliZ(0), qml.Hadamard(0) @ qml.PauliZ(1), [0], [0, 1], "mcm"]
     )
+    @pytest.mark.parametrize("use_kwarg", [True, False])
```
Comment (vincentmr) [inline frontend/test/pytest/test_mid_circuit_measurement.py:297]: I would expect this to be costly. Could we skip most tests for one of them?

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 2]
```diff
@@ -318,7 +375,10 @@ def cfun0():
     @pytest.mark.parametrize("shots", [5000])
     @pytest.mark.parametrize("postselect", [None, 0, 1])
     @pytest.mark.parametrize("reset", [False, True])
-    def test_dynamic_one_shot_multiple_measurements(self, backend, shots, postselect, reset):
+    @pytest.mark.parametrize("use_kwarg", [True, False])
```
Comment (vincentmr) [inline frontend/test/pytest/test_mid_circuit_measurement.py:378]: Same.

**Review by vincentmr — COMMENTED** [review]
I like that way of specifying dynamic circuit execution. Thanks @mudit2812 , look good and I just have a suggestion about the tests. Let me know what you think.

**Comment by dime10** [comment]
Collecting todos from Vincents PR:
- https://github.com/PennyLaneAI/catalyst/pull/773#discussion_r1631454249
- https://github.com/PennyLaneAI/catalyst/pull/773#discussion_r1631357408

File: frontend/test/pytest/test_mid_circuit_measurement.py (hunk 1 above)
Comment (mudit2812) [reply frontend/test/pytest/test_mid_circuit_measurement.py:297]: If we want to make the use of `mcm_method` the default way to request MCM transforms, then I'll update the tests to use that instead of decorating with `dynamic_one_shot`.

File: .dep-versions [hunk 3]
```diff
@@ -5,4 +5,5 @@ llvm=cd9a641613eddf25d4b25eaa96b2c393d401d42c
 enzyme=1beb98b51442d50652eaa3ffb9574f4720d611f1
 
 # Always remove custom PL/LQ versions before release.
-pennylane=7f97ea735b3c43b43a0ffe34040bffaa49603268
\ No newline at end of file
+# pennylane=7f97ea735b3c43b43a0ffe34040bffaa49603268
+pennylane=497d22ffa31be6e0213bb85b7809e6df441a15cc
```
Comment (mudit2812) [inline .dep-versions:9]: Once https://github.com/PennyLaneAI/pennylane/pull/5833 is merged.
```suggestion
pennylane=7f97ea735b3c43b43a0ffe34040bffaa49603268
```

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 4]
```diff
@@ -315,18 +314,30 @@ class MidCircuitMeasure(HybridOp):
...
+        in_classical_tracers,
+        out_classical_tracers,
+        regions: List[HybridOpRegion],
+        reset: bool = None,
+        postselect: int = None,
+    ):
+        HybridOp.__init__(self, in_classical_tracers, out_classical_tracers, regions)
+        self.reset = reset
+        self.postselect = postselect
+        # self.mv = self.out_classical_tracers[0]
```
Comment (vincentmr) [inline frontend/catalyst/api_extensions/quantum_operators.py:329]: What is the TODO going with that again?

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 5]
```diff
@@ -315,18 +314,30 @@ class MidCircuitMeasure(HybridOp):
...
 
-    def trace_quantum(self, ctx, device, trace, qrp) -> QRegPromise:
-        op = self
-        wire = op.in_classical_tracers[0]
+    # pylint: disable=too-many-arguments
+    def trace_quantum(self, ctx, device, trace, qrp, postselect_mode=None) -> QRegPromise:
+        wire = self.in_classical_tracers[0]
```
Comment (vincentmr) [inline frontend/catalyst/api_extensions/quantum_operators.py:333]: Good, I stumbled on the alias once or twice.

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 6]
```diff
@@ -315,18 +314,30 @@ class MidCircuitMeasure(HybridOp):
...
         qubit = qrp.extract([wire])[0]
-        # qubit2 = op.bind_overwrite_classical_tracers(ctx, trace, qubit)
-        # TODO: execute post-selection depending on qnode config
-        qubit2 = op.bind_overwrite_classical_tracers(ctx, trace, qubit, postselect=op.postselect)
+        if postselect_mode != "hw-like":
```
Comment (vincentmr) [inline frontend/catalyst/api_extensions/quantum_operators.py:335]: Nitpick, I prefer `==` conditions when possible

File: frontend/catalyst/autograph/ag_primitives.py [hunk 7]
```diff
@@ -271,7 +271,7 @@ def for_stmt(
     #   -> this will raise a warning to allow users to correct mistakes and allow the conversion
     #      to succeed, for example because they forgot to use a list instead of an array
 
-    # pylint: disable=multiple-statements,missing-class-docstring
+    # pylint: disable=missing-class-docstring
```
Comment (vincentmr) [inline frontend/catalyst/autograph/ag_primitives.py:274]: Seems like this was silencing a warning.
```suggestion
    # pylint: disable=multiple-statements,missing-class-docstring
```

File: frontend/catalyst/qfunc.py [hunk 8]
```diff
@@ -89,6 +89,42 @@ def extract_backend_info(
     def __call__(self, *args, **kwargs):
         assert isinstance(self, qml.QNode)
 
+        # Mid-circuit measurement configuration/execution
+        mcm_config = self.execute_kwargs["mcm_config"]
+        _dynamic_one_shot_called = kwargs.pop("_dynamic_one_shot_called", False)
+        total_shots = (
+            self.device.shots
+            if isinstance(self.device, qml.devices.LegacyDevice)
+            else self.device.shots.total_shots
```
Comment (vincentmr) [inline frontend/catalyst/qfunc.py:98]: Are there many places where this could be avoided implementing a `get_total_shots` method in all devices?

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 9]
```diff
@@ -218,15 +221,134 @@ def circuit(x):
...
+    def test_mcm_method_deferred_error(self, backend):
+        """Test that an error is raised if trying to execute with mcm_method="deferred"."""
+        dev = qml.device(backend, wires=1)
+
+        @qjit
+        @qml.qnode(dev, mcm_method="deferred")
+        def circuit(x):
+            qml.RY(x, wires=0)
+            m = measure(0)
+            qml.PauliX(0)
```
Comment (vincentmr) [inline frontend/test/pytest/test_mid_circuit_measurement.py:233]: For simplicity.
```suggestion
```

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 10]
```diff
@@ -218,15 +221,134 @@ def circuit(x):
...
+    def test_mcm_method_one_shot_analytic_error(self, backend):
+        """Test that an error is raised if using mcm_method="one-shot" without shots."""
+        dev = qml.device(backend, wires=1, shots=None)
+
+        @qjit
+        @qml.qnode(dev, mcm_method="one-shot")
+        def circuit(x):
+            qml.RY(x, wires=0)
+            m = measure(0)
+            qml.PauliX(0)
```
Comment (vincentmr) [inline frontend/test/pytest/test_mid_circuit_measurement.py:251]: Idem.
```suggestion
```

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 11]
```diff
@@ -218,15 +221,134 @@ def circuit(x):
...
+        """Test that an error is raised if using `mcm_method="single-branch-statistics"` and
+        `postselect_mode="hw-like"`"""
+        dev = qml.device(backend, wires=1, shots=10)
+
+        @qjit
+        @qml.qnode(dev, mcm_method=mcm_method, postselect_mode="hw-like")
+        def circuit(x):
+            qml.RY(x, wires=0)
+            m = measure(0)
+            qml.PauliX(0)
```
Comment (vincentmr) [inline frontend/test/pytest/test_mid_circuit_measurement.py:271]: Or maybe like that?
```suggestion
            m = measure(0)
```

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 12]
```diff
@@ -218,15 +221,134 @@ def circuit(x):
...
     def test_dynamic_one_shot_unsupported_measurement(self, backend):
         """Test that circuits with unsupported measurements raise an error."""
         shots = 10
         dev = qml.device(backend, wires=1, shots=shots)
         param = np.pi / 4
 
         @qjit
-        @catalyst.qfunc.dynamic_one_shot
-        @qml.qnode(dev)
+        @qml.qnode(dev, mcm_method="one-shot")
```
Comment (vincentmr) [inline frontend/test/pytest/test_mid_circuit_measurement.py:351]: Nice!

**Review by vincentmr — COMMENTED** [review]
Great work @mudit2812 ! Not much to raise on my end and pretty much ready to merge IMO.

File: frontend/catalyst/qfunc.py (hunk 8 above)
Comment (mudit2812) [reply frontend/catalyst/qfunc.py:98]: Probably. There are variations of the same logic across PL as well. I think I'll add a private helper to `qfunc.py` to get the shots and we can replace it if/when a similar helper is added to vanilla PL

File: .dep-versions [hunk 13]
```diff
@@ -5,4 +5,4 @@ llvm=cd9a641613eddf25d4b25eaa96b2c393d401d42c
 enzyme=1beb98b51442d50652eaa3ffb9574f4720d611f1
 
 # Always remove custom PL/LQ versions before release.
-pennylane=7f97ea735b3c43b43a0ffe34040bffaa49603268
\ No newline at end of file
+pennylane=4eb803abf19c2438b8d4d208525a727c294b2bbd
```
Comment (dime10) [inline .dep-versions:8]: This is a reminder for later to double check which hash is newer once changes from main are incorporated.

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 5 above)
Comment (dime10) [inline frontend/catalyst/api_extensions/quantum_operators.py:333]: Maybe we can assign it in the init as well?

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 6 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:335]: Yes especially in this case I think applying the post-select should be the default :)

File: frontend/catalyst/jax_primitives.py [hunk 14]
```diff
@@ -77,7 +77,7 @@
 from catalyst.utils.extra_bindings import FromElementsOp, TensorExtractOp
 from catalyst.utils.types import convert_shaped_arrays_to_tensors
 
-# pylint: disable=unused-argument,abstract-method,too-many-lines
+# pylint: disable=unused-argument,too-many-lines
```
Comment (dime10) [inline frontend/catalyst/jax_primitives.py:80]:
```suggestion
# pylint: disable=unused-argument,abstract-method,too-many-lines
```
I think this is important unfortunately, have definitely seen that warning floating around. Super annoying that they don't appear to be very consistent :/

File: frontend/catalyst/jax_tracer.py [hunk 15]
```diff
@@ -400,16 +400,19 @@ def lower_jaxpr_to_mlir(jaxpr, func_name):
...
+# pylint: disable=too-many-arguments
 @debug_logger
-def trace_quantum_tape(
+def trace_quantum_operations(
     quantum_tape: QuantumTape,
     device: QubitDevice,
     qreg: DynamicJaxprTracer,
     ctx: JaxTracingContext,
     trace: DynamicJaxprTrace,
+    qnode: qml.QNode = None,
```
Comment (dime10) [inline frontend/catalyst/jax_tracer.py:411]: Should the qnode ever be none?

File: frontend/catalyst/jax_tracer.py [hunk 16]
```diff
@@ -486,12 +489,18 @@ def _bind_native_controlled_op(qrp, op, controlled_wires, controlled_values):
...
+            kwargs = (
+                {
+                    "postselect_mode": getattr(qnode, "execute_kwargs", {})
+                    .get("mcm_config", qml.devices.MCMConfig())
+                    .postselect_mode
+                }
+                if isinstance(op, catalyst.api_extensions.quantum_operators.MidCircuitMeasure)
+                else {}
+            )
+            qrp2 = op.trace_quantum(ctx, device, trace, qrp, **kwargs)
```
Comment (dime10) [inline frontend/catalyst/jax_tracer.py:501]: Would it be simpler to just pass in the execution or mcm config? Fairly minor admittedly but it could make the signature more uniform.

File: frontend/catalyst/qfunc.py [hunk 17]
```diff
@@ -89,6 +89,42 @@ def extract_backend_info(
     def __call__(self, *args, **kwargs):
         assert isinstance(self, qml.QNode)
 
+        # Mid-circuit measurement configuration/execution
+        mcm_config = self.execute_kwargs["mcm_config"]
+        _dynamic_one_shot_called = kwargs.pop("_dynamic_one_shot_called", False)
```
Comment (dime10) [inline frontend/catalyst/qfunc.py:94]: I don't know that I like this being passes as a call argument to the qnode. I would prefer if it was a property in that case. Or maybe it's a special shot value? Or a subclass of QNode?

File: frontend/catalyst/qfunc.py [hunk 18]
```diff
@@ -121,7 +157,7 @@ def _eval_quantum(*args):
         return tree_unflatten(out_tree_promise(), res_flat)[0]
 
 
-# pylint: disable=protected-access,no-member,not-callable
+# pylint: disable=not-callable
```
Comment (dime10) [inline frontend/catalyst/qfunc.py:160]: This is wild, didn't vincent just add these? Truly a headache 😭

File: frontend/catalyst/third_party/cuda/primitives/__init__.py [hunk 19]
```diff
@@ -24,7 +24,7 @@
 
 # We disable protected access in particular to avoid warnings with cudaq._pycuda.
 # And we disable unused-argument to avoid unused arguments in abstract_eval, particularly kwargs.
-# pylint: disable=protected-access,unused-argument,abstract-method,line-too-long
+# pylint: disable=protected-access,unused-argument,line-too-long
```
Comment (dime10) [inline frontend/catalyst/third_party/cuda/primitives/__init__.py:27]: Same as earlier
```suggestion
# pylint: disable=protected-access,unused-argument,abstract-method,line-too-long
```

**Review by dime10 — COMMENTED** [review]
Thanks @mudit2812, nice work!! 🚀 Nothing major really

> Setting postselect_mode="hw-like" without specifying mcm_method="one-shot" will fail. This includes when no value for mcm_method is specified.

Any reason not to set the mcm method automatically?

File: frontend/catalyst/autograph/ag_primitives.py (hunk 7 above)
Comment (mudit2812) [reply frontend/catalyst/autograph/ag_primitives.py:274]: I was getting a codefactor error that said that I was silencing a pylint warning that didn't need to be silenced, which is why I removed it.

Comment (mudit2812) [reply frontend/catalyst/autograph/ag_primitives.py:274]: Very odd that codefactor is still complaining (especially because pylint is not complaning).

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 4 above)
Comment (mudit2812) [reply frontend/catalyst/api_extensions/quantum_operators.py:329]: Sorry I'm not sure I understand the question.

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 5 above)
Comment (mudit2812) [reply frontend/catalyst/api_extensions/quantum_operators.py:333]: We would need a way for `catalyst.measure` to know the `postselect_mode` for that.

File: frontend/catalyst/jax_tracer.py (hunk 15 above)
Comment (mudit2812) [reply frontend/catalyst/jax_tracer.py:411]: Not as far as I know. I will make it a positional argument.

File: frontend/catalyst/qfunc.py (hunk 17 above)
Comment (mudit2812) [reply frontend/catalyst/qfunc.py:94]: I've made them temporary attributes of the qnode that get deleted at the end of the execution. Let me know what you think about that implementation.

File: frontend/catalyst/qfunc.py (hunk 18 above)
Comment (mudit2812) [reply frontend/catalyst/qfunc.py:160]: I'll add them back and see if CodeFactor complains 😓

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 4 above)
Comment (vincentmr) [reply frontend/catalyst/api_extensions/quantum_operators.py:329]: Do we want to simply delete that line, or use it in the future? If so, what's the intent behind this line?

Comment (mudit2812) [reply frontend/catalyst/api_extensions/quantum_operators.py:329]: Oh I see. I'm going to remove that line as it's not related to the changes in this PR. I was prototyping the `MidMeasureMP` inheritance and forgot to remove it.

**Comment by mudit2812** [comment]
> Thanks @mudit2812, nice work!! 🚀 Nothing major really
> 
> > Setting postselect_mode="hw-like" without specifying mcm_method="one-shot" will fail. This includes when no value for mcm_method is specified.
> 
> Any reason not to set the mcm method automatically?

Fixed

**Comment by mudit2812** [comment]
Getting a bunch of useless suppression warnings [here](https://github.com/PennyLaneAI/catalyst/actions/runs/9506271437/job/26203073536?pr=798)

File: frontend/catalyst/qfunc.py [hunk 20]
```diff
@@ -206,13 +247,18 @@ def processing_fn(results):
...
-    single_name = dev.short_name if isinstance(dev, qml.devices.LegacyDevice) else dev.name
-    single_shot_qnode.device = qml.device(single_name, wires=dev.wires, shots=single_shot)
-    total_shots = (
-        dev.shots if isinstance(qnode.device, qml.devices.LegacyDevice) else dev.shots.total_shots
-    )
+    total_shots = _get_device_shots(dev)
+
+    new_dev = copy(dev)
+    if isinstance(new_dev, qml.devices.LegacyDevice):
+        new_dev.shots = 1
```
Comment (mudit2812) [inline frontend/catalyst/qfunc.py:258]: Codecov says that this line is not covered, but it should be covered since the frontend tests use `lightning.kokkos` as well. Anyone have any ideas why this might be happening?

File: frontend/catalyst/jax_tracer.py (hunk 16 above)
Comment (mudit2812) [reply frontend/catalyst/jax_tracer.py]: I updated it to take `mcm_config` as an optional kwarg.

File: frontend/catalyst/qfunc.py (hunk 18 above)
Comment (mudit2812) [reply frontend/catalyst/qfunc.py]: The formatting check failed due to useless suppressions. See [here](https://github.com/PennyLaneAI/catalyst/actions/runs/9506271437/job/26203073536?pr=798).

File: frontend/catalyst/qfunc.py [hunk 21]
```diff
@@ -62,6 +63,11 @@
 logger.addHandler(logging.NullHandler())
 
 
+def _get_device_shots(dev):
```
Comment (vincentmr) [inline frontend/catalyst/qfunc.py:66]: Can we use this [here](https://github.com/PennyLaneAI/catalyst/blob/mcm-kwargs/frontend/catalyst/jax_tracer.py#L661) and [here](https://github.com/PennyLaneAI/catalyst/blob/mcm-kwargs/frontend/catalyst/device/qjit_device.py#L149)?

File: frontend/catalyst/qfunc.py (hunk 20 above)
Comment (vincentmr) [reply frontend/catalyst/qfunc.py:258]: Looks good now, doesn't it?

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 22]
```diff
@@ -218,15 +222,185 @@ def circuit(x):
...
+    def test_mcm_method_deferred_error(self, backend):
+        """Test that an error is raised if trying to execute with mcm_method="deferred"."""
+        dev = qml.device(backend, wires=1)
+
+        @qjit
+        @qml.qnode(dev, mcm_method="deferred")
+        def circuit(_x):  # pylint: disable=unused-argument
+            # QNode has unused argument because if there are no arguments then compilation happens
+            # during QJIT.__init__ rather than __call__
+            m = measure(0)
```
Comment (vincentmr) [inline frontend/test/pytest/test_mid_circuit_measurement.py:234]: This is unrelated to the above right?
```suggestion
            _ = measure(0)
```

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 23]
```diff
@@ -218,15 +222,185 @@ def circuit(x):
...
+    def test_mcm_method_one_shot_analytic_error(self, backend):
+        """Test that an error is raised if using mcm_method="one-shot" without shots."""
+        dev = qml.device(backend, wires=1, shots=None)
+
+        @qjit
+        @qml.qnode(dev, mcm_method="one-shot")
+        def circuit(_x):  # pylint: disable=unused-argument
+            # QNode has unused argument because if there are no arguments then compilation happens
+            # during QJIT.__init__ rather than __call__
+            m = measure(0)
```
Comment (vincentmr) [inline frontend/test/pytest/test_mid_circuit_measurement.py:251]:
```suggestion
            _ = measure(0)
```

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 24]
```diff
@@ -218,15 +222,185 @@ def circuit(x):
...
+        """Test that an error is raised if using `mcm_method="single-branch-statistics"` and
+        `postselect_mode="hw-like"`"""
+        dev = qml.device(backend, wires=1, shots=10)
+
+        @qjit
+        @qml.qnode(dev, mcm_method="single-branch-statistics", postselect_mode="hw-like")
+        def circuit(_x):  # pylint: disable=unused-argument
+            # QNode has unused argument because if there are no arguments then compilation happens
+            # during QJIT.__init__ rather than __call__
+            m = measure(0)
```
Comment (vincentmr) [inline frontend/test/pytest/test_mid_circuit_measurement.py:269]:
```suggestion
            _ = measure(0)
```

**Review by vincentmr — APPROVED** [review]
Looks good to me, happy to approve! I'm leaving a suggestion to use `_get_device_shots` in a couple places.

File: frontend/catalyst/qfunc.py (hunk 20 above)
Comment (mudit2812) [reply frontend/catalyst/qfunc.py:258]: Sadly not. I may have to add a `# pragma: no cover`

**Comment by mudit2812** [comment]
> Looks good to me, happy to approve! I'm leaving a suggestion to use `_get_device_shots` in a couple places.

I moved it to `jax_tracer.py` and made it a public function so that other files can import and use it. @dime10 let me know if it makes sense to move the helper function to a different file.

File: frontend/catalyst/qfunc.py [hunk 25]
```diff
@@ -202,13 +237,18 @@ def processing_fn(results):
...
-    single_name = dev.short_name if isinstance(dev, qml.devices.LegacyDevice) else dev.name
-    single_shot_qnode.device = qml.device(single_name, wires=dev.wires, shots=single_shot)
-    total_shots = (
-        dev.shots if isinstance(qnode.device, qml.devices.LegacyDevice) else dev.shots.total_shots
-    )
+    total_shots = get_device_shots(dev)
+
+    new_dev = copy(dev)
+    if isinstance(new_dev, qml.devices.LegacyDevice):
+        new_dev.shots = 1  # pragma: no cover
```
Comment (mudit2812) [inline frontend/catalyst/qfunc.py:248]: This is the uncovered line, so I've added a `# pragma: no cover` for now.

File: frontend/catalyst/qfunc.py [hunk 26]
```diff
@@ -202,13 +237,18 @@ def processing_fn(results):
...
+    else:
+        new_dev._shots = qml.measurements.Shots(1)
+    single_shot_qnode.device = new_dev
```
Comment (mudit2812) [inline frontend/catalyst/qfunc.py:251]: This is the uncovered line, so I've added a `# pragma: no cover` for now. I could also fix this by doing the following, but it is much less readable and honestly just bad.

```suggestion
    new_shots = 1 if instance(new_dev, qml.devices.LegacyDevice) else qml.measurements.Shots(1)
    setattr(new_dev, "shots" if isinstance(new_dev, qml.devices.LegacyDevice) else "_shots", new_shots)
```

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 5 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:335]: For the wire? I just meant parsing out the wire on init from the classical tracers, mainly for "nicecess" but very minor.

File: frontend/catalyst/qfunc.py (hunk 18 above)
Comment (dime10) [reply frontend/catalyst/qfunc.py]: Formatting/linting should never fail due to the useless suppression warning, since it is just "informational" (IXXX). In that case it was and actual linting error that failed the check (E1101).

File: frontend/catalyst/qfunc.py (hunk 20 above)
Comment (dime10) [reply frontend/catalyst/qfunc.py]: It's not blocking check so it's possible to merge without as well if you think the line should be covered but for some reason isn't. `# pragma: nocover` should ideally be used for lines that we explicitly don't want to cover.

File: frontend/catalyst/autograph/ag_primitives.py (hunk 7 above)
Comment (dime10) [reply frontend/catalyst/autograph/ag_primitives.py:274]: Can you restore the silencing? :)

File: frontend/catalyst/device/qjit_device.py [hunk 27]
```diff
@@ -35,6 +35,7 @@
     catalyst_decompose,
     measurements_from_counts,
 )
+from catalyst.jax_tracer import get_device_shots
```
Comment (dime10) [inline frontend/catalyst/device/qjit_device.py:38]: It's not clear to me immediately why `get_device_shots` is something that should be defined in the `jax_tracer` module, rather than something relating to devices 🤔

File: frontend/catalyst/jax_primitives.py [hunk 28]
```diff
@@ -83,7 +83,7 @@
 from catalyst.utils.extra_bindings import FromElementsOp, TensorExtractOp
 from catalyst.utils.types import convert_shaped_arrays_to_tensors
 
-# pylint: disable=unused-argument,abstract-method,too-many-lines,too-many-statements
+# pylint: disable=unused-argument,too-many-lines,too-many-statements
```
Comment (dime10) [inline frontend/catalyst/jax_primitives.py:86]: Let's leave existing silencing in place until we can guarantee it won't cause failures:
```suggestion
# pylint: disable=unused-argument,abstract-method,too-many-lines,too-many-statements
```

File: frontend/catalyst/qfunc.py (hunk 26 above)
Comment (dime10) [reply frontend/catalyst/qfunc.py:251]: I think it makes sense that it's uncovered since we perform code coverage with a legacy device anymore since lightning upgraded.

File: frontend/catalyst/qfunc.py [hunk 29]
```diff
@@ -87,6 +87,40 @@ def extract_backend_info(
...
+            if mcm_config.mcm_method == "deferred":
+                raise ValueError("mcm_method='deferred' is not supported with Catalyst.")
+            if (
+                mcm_config.mcm_method == "single-branch-statistics"
+                and mcm_config.postselect_mode == "hw-like"
+            ):
+                raise ValueError(
+                    "Cannot use postselect_mode='hw-like' with Catalyst when "
+                    "mcm_method != 'one-shot'."
+                )
```
Comment (dime10) [inline frontend/catalyst/qfunc.py:113]: Should all the validation be a separate function? The if block is a bit long so I didn't even realize there was an additional return in there at the end.

It could be a function that runs all the checks and then returns the `mcm_config`. Then the conditional would be something like:
```py
if not dynamic_one_shot_called:
    mcm_config = self.validate_mcm_config()
    if mcm_config.mcm_method == "one-shot":
        mcm_config.postselect_mode = mcm_config.postselect_mode or "hw-like"
        return dynamic_one_shot(self, mcm_config=mcm_config)(*args, **kwargs)
```

File: frontend/catalyst/jax_tracer.py [hunk 30]
```diff
@@ -1073,8 +1080,11 @@ def is_leaf(obj):
                     output = return_values_flat
                     trees = return_values_tree
 
-                qrp_out = trace_quantum_tape(tape, device, qreg_in, ctx, trace)
-                meas, meas_trees = trace_quantum_measurements(device, qrp_out, output, trees, tape)
+                mcm_config = (
+                    getattr(qnode, "_tmp_mcm_config", None) or qnode.execute_kwargs["mcm_config"]
+                )
```
Comment (dime10) [inline frontend/catalyst/jax_tracer.py:1085]: I'm confused, is `_tmp_mcm_config` set anywhere? If so I couldn't find it 😅 Also thought it wouldn't be required anymore since the one shot QNode can be freely mutated without affecting the user qnode.

**Review by dime10 — APPROVED** [review]
Thanks @mudit2812, I'm happy to approve with some minor fixups (also previous open comments) :)

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 5 above)
Comment (mudit2812) [reply frontend/catalyst/api_extensions/quantum_operators.py:335]: Oh sorry I got confused. Thought you meant the `postselect_mode`.

File: frontend/catalyst/jax_tracer.py (hunk 30 above)
Comment (mudit2812) [reply frontend/catalyst/jax_tracer.py:1085]: I think I may have forgotten to remove this after I changed the implementation.

File: frontend/catalyst/device/qjit_device.py (hunk 27 above)
Comment (mudit2812) [reply frontend/catalyst/device/qjit_device.py:38]: It was mostly to avoid circular dependencies. I'll move it to `qjit_device.py` and if that works I'll keep it there.

Comment (mudit2812) [reply frontend/catalyst/device/qjit_device.py:38]: There was a circular dependency, so I duplicated the helper in `jax_tracer.py` as well.
